link time error: multiple definition of method

This code:

```cr
def a(x)
  x
end

class A
  def a(x)
    super x
  end
end

a "1"
A.new.a "2"
```

Causes:

```
A-.o: In function `a':
/home/guilherme/Cubos/sdkgen2/x.cr:3: multiple definition of `*a<String>:String'
_main.o:/home/guilherme/Cubos/sdkgen2/x.cr:3: first defined here
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc "${@}" -o '/home/guilherme/.cache/crystal/crystal-run-x.tmp'  -rdynamic  -lpcre -lgc -lpthread /opt/crystal/src/ext/libcrystal.a -levent -lrt -ldl -L/usr/lib -L/usr/local/lib`
```

This happens on both lastest release (Crystal 0.23.1 [e2a1389e8] (2017-07-13) LLVM 3.8.1) and current master (Crystal 0.24.0 [ea4187c57] (2017-10-28) LLVM 4.0.1).

Obviously the code makes no sense, there is no "super" to invoke. But it surely should report correct error. This is my expected output:

```
Error in x.cr:12: instantiating 'A#a(String)'

A.new.a "2"
      ^

in x.cr:7: undefined method 'a'

    super x
    ^~~~~
```

--- a/spec/compiler/semantic/super_spec.cr
+++ b/spec/compiler/semantic/super_spec.cr
@@ -366,4 +366,20 @@ describe "Semantic: super" do
       ),
       "can't use 'super' outside method"
   end
+
+  it "errors on super where only target would be a top level method (#5201)" do
+    assert_error %(
+      def bar
+      end
+
+      class Foo
+        def bar
+          super
+        end
+      end
+
+      Foo.new.bar
+      ),
+      "no overload matches 'bar'"
+  end
 end

--- a/src/compiler/crystal/semantic/call.cr
+++ b/src/compiler/crystal/semantic/call.cr
@@ -217,7 +217,7 @@ class Crystal::Call
     instantiate matches, owner, self_type: nil
   end
 
-  def lookup_matches_in_type(owner, arg_types, named_args_types, self_type, def_name, search_in_parents)
+  def lookup_matches_in_type(owner, arg_types, named_args_types, self_type, def_name, search_in_parents, search_in_toplevel = true)
     signature = CallSignature.new(def_name, arg_types, block, named_args_types)
 
     matches = check_tuple_indexer(owner, def_name, args, arg_types)
@@ -225,7 +225,7 @@ class Crystal::Call
 
     # If we didn't find a match and this call doesn't have a receiver,
     # and we are not at the top level, let's try searching the top-level
-    if matches.empty? && !obj && owner != program
+    if matches.empty? && !obj && owner != program && search_in_toplevel
       program_matches = lookup_matches_with_signature(program, signature, search_in_parents)
       matches = program_matches unless program_matches.empty?
     end
@@ -592,10 +592,10 @@ class Crystal::Call
     if parents && parents.size > 0
       parents.each_with_index do |parent, i|
         if parent.lookup_first_def(enclosing_def.name, block)
-          return lookup_matches_in_type(parent, arg_types, named_args_types, scope, enclosing_def.name, !in_initialize)
+          return lookup_matches_in_type(parent, arg_types, named_args_types, scope, enclosing_def.name, !in_initialize, search_in_toplevel: false)
         end
       end
-      lookup_matches_in_type(parents.last, arg_types, named_args_types, scope, enclosing_def.name, !in_initialize)
+      lookup_matches_in_type(parents.last, arg_types, named_args_types, scope, enclosing_def.name, !in_initialize, search_in_toplevel: false)
     else
       raise "there's no superclass in this scope"
     end

